backend/cellguide/pipeline/ontology_tree/tree_builder.py

```diff
@@ -0,0 +1,188 @@
+The tree is populated with cell type counts from an input dataframe.
+"""
+
+
+class OntologyTreeBuilder:
+    def __init__(self, cell_counts_df, root_node="CL:0000548"):
+        """
+        OntologyTreeBuilder is a class that builds a nested dictionary representation of the cell ontology tree
+        and populates the nodes with cell type counts from the input cell counts dataframe.
+
+        Arguments
+        ---------
+        cell_counts_df: pandas.DataFrame
+            A dataframe with required columns "tissue_ontology_term_id", "cell_type_ontology_term_id", and "n_cells".
+            This dataframe is typically the cell counts cube from the WMG snapshot.
+
+        root_node: str, optional, default="CL:0000548" (animal cell)
+            The root node of the ontology tree. This is the node from which the ontology tree is traversed.
+        """
+
+        logger.info("Loading CL ontology...")
+        self.ontology = Ontology(CL_BASIC_PERMANENT_URL_PRONTO)
+
+        logger.info("Initializing cell type data structures from the input cell counts dataframe...")
+        self.id_to_name, self.all_cell_type_ids = self._initialize_id_to_name()
+        self.cell_counts_df, self.cell_counts_df_rollup = self._initialize_cell_counts_df_rollup(cell_counts_df)
+        self.all_cell_type_ids_in_corpus = self.cell_counts_df_rollup.index.values[
+            self.cell_counts_df_rollup.values > 0
+        ]
+        logger.info("Initializing ontology tree data structures by traversing CL ontology...")
+        self.ontology_graph, self.traverse_node_counter, self.all_unique_nodes = self._traverse_ontology_with_counting(
+            self.ontology[root_node]
+        )
+        self._delete_unknown_terms_from_ontology_graph(self.ontology_graph)
+        self.start_node = f"{root_node}__0"
+
+    def _initialize_cell_counts_df_rollup(self, cell_counts_df):
+        """
+        This function prepares the cell counts dataframe and its rollup for the ontology tree traversal.
+
+        Arguments
+        ---------
+        cell_counts_df: pandas.DataFrame
+            This dataframe is typically the cell counts cube from the WMG snapshot.
+
+        Returns
+        -------
+        cell_counts_df: pandas.Series
+            A series with index "cell_type_ontology_term_id" and values "n_cells".
+            This series is the cell counts dataframe with the following modification:
+            1. The cell type ontology term ids that are not in the input cell counts dataframe are added with 0 counts.
+
+        cell_counts_df_rollup: pandas.Series
+            A series with index "cell_type_ontology_term_id" and values "n_cells".
+            "n_cells" contains the sum of "n_cells" for the cell type and its descendants.
+        """
+
+        cell_counts_df = (
+            cell_counts_df.groupby("cell_type_ontology_term_id").sum(numeric_only=True)[["n_cells"]].reset_index()
```

Comment:
just for my own understanding, what does `reset_index` here do?

Reply:
`groupby(...).sum(...)` converts the group-by dimensions into a MultiIndex (where each index value is a tuple of the group values). In the case where the group-by is one dimension, then it just converts the groupby dimension into an Index (each index value is a cell type in this case). 

`reset_index` converts the Index/MultiIndex back into the column/columns.